tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -274,6 +368,65 @@ std::vector<TensorView*> mha(
   return {qkv, sdpa_output, linear2, dropout_result};
 }
 
+// TODO: These linear_backwards helper functions can be merged once
+// we do not have logically split rfactor domain.
+struct LinearBackwardsResult {
+  TensorView* grad_x;
+  TensorView* grad_w;
+  TensorView* grad_b;
+};
+
+// x format: [i0, i1] dtype
+// weight format: [DID(D), i1, i2/D] dtype
+// grad format: [DID(D) i0, i2/D] float
+// outputs: grad_x [i0, i1] dtype
+// grad_w [DID i1, i2/D] dtype
+// grad_b [DID i2/2] float
+LinearBackwardsResult linear_backwards(
+    TensorView* x,
+    TensorView* w,
+    TensorView* grad,
+    DataType dtype) {
+  TensorView* grad_q;
+  if (grad->dtype() == dtype) {
+    grad_q = grad;
+    grad = castOp(DataType::Float, grad_q);
+  } else {
+    grad_q = castOp(dtype, grad);
+  }
```

Comment:
```suggestion
  TensorView* grad_q = maybeCastOp(dtype, grad);
```
and replace `sum(grad, ...)` later with `sum(maybeCastOp(Float, grad))`.

Reply:
Btw, can `grad` ever be lower than float? Apparently, `mlp` and `reference_mlp` both output float.

Reply:
The incoming grads get quantized to half/bfloat16 for the backwards linear operations since the weights and bias are in lower precision.

Reply:
I understood that -- that's what `grad_q` is for. But `grad` itself, the input of this function, is always float. Correct? If so, you can simply say `sum(grad)` somewhere below.

Reply:
No, it depends on the backwards operator that generates the incoming grad.
For MLP the incoming grad is float since it was generated from gelu.
For MHA the incoming grad is half since it was generated from sdpa.

Reply:
Got it!

Reply:
I don't think this has been resolved yet. Even though `grad`'s dtype is not fixed, we can still simplify the code using maybeCastOp as mentioned in the first comment in this thread.

Reply:
sure i can change it to use two maybes instead...